sklearn/metrics/_regression.py

```diff
@@ -855,9 +870,17 @@ def median_absolute_error(
 
 
 def _assemble_r2_explained_variance(
-    numerator, denominator, n_outputs, multioutput, force_finite
+    numerator, denominator, n_outputs, multioutput, force_finite, xp=None
 ):
     """Common part used by explained variance score and :math:`R^2` score."""
+    if xp is None:
+        input_arrays = [numerator, denominator]
+        if multioutput is not None and not isinstance(multioutput, str):
+            input_arrays.append(multioutput)
+        xp, _ = get_namespace(*input_arrays)
+
+    device_ = device(numerator)
+    dtype = numerator.dtype
```

Comment:
This seems like a repeated boilerplate that I've seen a lot in these PRs, can we not reduce the diff by having them in a function or something?

Reply:
You mean "get the device and the dtype, store them in a variable and use them later"?

The reason I'm asking for clarification is that I agree that it feels like we need quite some boilerplate when we add array API support, but there is so many variations of it. Hence asking what you mean :D

---

I think this change is the in `r2_score` PR already, so unrelated to "fixing `train_test_split`". Maybe worth raising in https://github.com/scikit-learn/scikit-learn/pull/27904